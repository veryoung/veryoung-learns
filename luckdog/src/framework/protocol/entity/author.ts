/** author实体 */
export class Author {
  /** 作者名 */
  public authorName = '';
  /** 作者标识 */
  public authorId = '';
  /** 作者相关图片 */
  public picUrl = '';
}
