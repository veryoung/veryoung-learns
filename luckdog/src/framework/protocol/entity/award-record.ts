/** 福利卡片中奖信息 */
export class AwardRecord {
  /** 头像 */
  public avatar = '';
  /** 中奖信息 */
  public awardDesc = '';
}
