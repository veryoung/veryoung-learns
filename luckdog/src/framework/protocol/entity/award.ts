/** 福利卡片奖品类型 */
export enum AwardType {
  /** 红包 */
  RED_PACK = 'RED_PACK',
  /** 快乐金币 */
  COIN = 'COIN',
  /** 腾讯视频会员 */
  VIDEO_VIP = 'VIDEO_VIP',
  /** 王者礼包 */
  LOTTERY_PACKAGE = 'LOTTERY_PACKAGE',
}

/** 福利卡片奖品 */
export class Award {
  /** 奖品图片 */
  public imageUrl = '';
  /** 奖品名称 */
  public name = '';
  /** 奖品类型 */
  public type: AwardType = AwardType.COIN;
}
