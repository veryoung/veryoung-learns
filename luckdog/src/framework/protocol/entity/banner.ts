/** banner实体 */
export class Banner {
  /** banner对应跳转地址 */
  public jumpUrl = '';
  /** banner图 */
  public bannerUrl = '';
}
