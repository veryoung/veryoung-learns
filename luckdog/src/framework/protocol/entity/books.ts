import { Tag } from './tag';

/** 书籍实体的基类 */
export class BaseBook {
  /** 资源id */
  public resourceId = '';
  /** 资源名称 */
  public resourceName = '';
  /** 简介 */
  public brief?: string;
  /** 作者 */
  public author?: string;
  /** 图片地址 */
  public picUrl?: string;
  /** 跳转地址 */
  public jumpUrl?: string;
  /** 推荐理由 */
  public reason?: string[];
  /** 分数 */
  public score?: string;
  /** 人气 (这里是处理好之后的值,携带对应数值单位)*/
  public hot?: string;
  /** 分类 (xx,yy,zz) */
  public category?: string;
  /** 书籍状态 */
  public status?: string;
  /** 书籍属性标签(独家 原创标签)*/
  public tag?: Tag;
  /** 书籍状态标签(完本 更新 连载) */
  public statusTag?: Tag;
  /** 运营语 */
  public description?: string;
}

