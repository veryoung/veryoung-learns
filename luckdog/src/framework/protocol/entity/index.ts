
export * from './author';
export * from './banner';
export * from './book-shelf';
export * from './books';
export * from './catch-up';
export * from './category';
export * from './kol';
export * from './nav';
export * from './tag';
export * from './video';
export * from './rank';
export * from './award';
export * from './award-record';
export * from './topic';
