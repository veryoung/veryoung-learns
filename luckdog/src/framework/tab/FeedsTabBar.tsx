/* eslint-disable react/sort-comp */
import React, { Component } from 'react';
import {
  Image,
  Text,
  View,
  Platform,
} from '@tencent/hippy-react-qb';
import FeedsTheme from '../FeedsTheme';
import FeedsProtect from '../../mixins/FeedsProtect';
import FeedsIcon from '../FeedsIcon';
import FeedsUtils from '../FeedsUtils';
import { DEFAULT_FEEDS_STYLE, EXT_TABBAR_CONFIG, TABLIST_CONFIG } from '../FeedsDefaultStyle';
import { dtConst } from '../FeedsConst';
import { CommonProps, DynamicTabItem, SkinModelType, TabIconItem, TabIconType, TabId } from '../../entity';
import { getTabpageBackgroundManager } from '@/presenters';
import { addKeylink, KeylinkCmd, logError } from '@/luckdog';
import { getDeviceVisitor, throttle } from '@/luckbox';

const TAG = 'FeedsTabBar';
/** 预加载节流频控 */
const PREFETCH_STEP = 3000;

const iconPrefetchedTabs: TabId[] = [];

interface Props extends CommonProps {
  tabInfo: DynamicTabItem;
  curTabId: TabId;
  tabsCount: number;
  left?: boolean;
  right?: boolean;
  redDotNum?: number;
}

type State = {
  /** 是否渲染图片 */
  useImage: boolean;
};

const getItemWidth = (
  tabTitle: string,
  tabParam: Record<string, any>,
  isSelected: boolean,
  tabIcons: TabIconItem[],
) => {
  let itemWidth;
  if (tabTitle.length <= 2) {
    // NBA三个字用两个汉字的宽度
    if (isSelected) {
      itemWidth = tabParam.focusWidth2;
    } else {
      itemWidth = tabParam.width2;
    }
  } else if (tabTitle.length === 3) {
    if (isSelected) {
      itemWidth = tabParam.focusWidth3;
    } else {
      itemWidth = tabParam.width3;
    }
  } else if (isSelected) {
    // 4个字及以上
    itemWidth = tabParam.focusWidth4;
  } else {
    itemWidth = tabParam.width4;
  }

  if (!tabIcons?.length) return itemWidth;

  const [tabIcon] = tabIcons;
  const { normalIconUrl, selectedIconUrl } = tabIcon;
  const iconType = tabIcon.iconType || TabIconType.UNKNOWN;
  const tabIconTypeWidth = EXT_TABBAR_CONFIG.tabIconType[iconType];
  const tabIconFocusTypeWidth = EXT_TABBAR_CONFIG.tabIconFocusType[iconType];
  if (normalIconUrl && !isSelected && tabIconTypeWidth) {
    itemWidth = tabIconTypeWidth;
  } else if (selectedIconUrl && isSelected && tabIconFocusTypeWidth) {
    itemWidth = tabIconFocusTypeWidth;
  }

  return itemWidth;
};

@FeedsProtect.protect
export default class NavItem extends Component<Props, State> {
  public state = {
    useImage: getDeviceVisitor().isAdr() ? this.props.tabInfo?.tabIcons?.length > 0 : false,
  };
  private prefetchSuccess = false;
  private prefetchTimer: NodeJS.Timeout | null = null;
  private prefetchTabIcon = throttle(async (tabInfo: DynamicTabItem): Promise<void> => {
    const { tabIcons, tabId } = tabInfo || {};
    if (!tabIcons?.length) return;

    // 安卓机在即便有图片缓存的情况下，预加载图片依然有一定耗时
    // 此时做一个 300ms 的定时器，如果 300ms 还未加载完成，就先切换到打底文字 tab
    const isAndroid = getDeviceVisitor().isAdr();
    if (isAndroid) {
      if (this.prefetchTimer) clearTimeout(this.prefetchTimer);
      this.prefetchTimer = setTimeout(() => {
        // 如果预加载已经提前完成，则不需要再切换到打底文字 tab 了
        if (this.prefetchSuccess) return;

        this.setState({
          useImage: false,
        });
      }, 300);
    }

    const start = Date.now();

    // 提取 icon url 进行预加载
    const images = tabIcons.reduce((acc: string[], { scene, normalIconUrl, selectedIconUrl }) => {
      // 非当前皮肤模式下的图片不做预加载
      if (scene as unknown as SkinModelType !== FeedsTheme.skinMode) return acc;
      return [...acc, normalIconUrl, selectedIconUrl];
    }, []).filter(url => !!url);

    addKeylink(`${tabId}-tabicon-prefetch`, KeylinkCmd.PR_INFO_SUM);

    // 不用加 try catch
    const prefetchStatus = await this.prefetchImages(images);

    // 针对安卓机做一个预加载图片平均耗时特性上报
    if (isAndroid) {
      addKeylink(`${tabId}-tabicon-prefetchtime=${Date.now() - start}`, KeylinkCmd.PR_AVG);
    }

    // 所有 icon 都加载成功才算成功
    const isAllSuccess = prefetchStatus.every(status => !!status);
    if (!isAllSuccess) return;

    this.onPrefetchSuccess(tabId);
  }, PREFETCH_STEP);

  public constructor(props) {
    super(props);
    const { tabInfo } = this.props;
    this.prefetchTabIcon(tabInfo);
  }

  public UNSAFE_componentWillReceiveProps(nextProps) {
    const tabID = nextProps.tabInfo.tabId || 0;
    if (!iconPrefetchedTabs.includes(tabID)) {
      this.prefetchTabIcon(nextProps.tabInfo);
    }
  }

  public render() {
    const { curTabId, left, right, tabInfo } = this.props;
    const styles = DEFAULT_FEEDS_STYLE.data.tabList;
    const extStyles = DEFAULT_FEEDS_STYLE.data.extTabList;
    const { tabParam: { leftPadding, paddingLeft, rightPadding, paddingRight } } = TABLIST_CONFIG;

    const itemWidth = getItemWidth(
      tabInfo.tabTitle,
      TABLIST_CONFIG.tabParam,
      tabInfo.tabId === curTabId,
      tabInfo.tabIcons,
    );

    const styleObj: Record<string, any> = {
      paddingLeft: left ? leftPadding : paddingLeft,
      paddingRight: right ? rightPadding : paddingRight,
    };

    // ADR text paddingBottom 有bug，设了paddingBottom反而往下跑了
    const focusView = tabInfo.tabId === curTabId ? styles.focusView : null;

    // 有异形图的情况，遵守短图（1~2字）与长图（3~4)字的两种宽度
    return (
      <View
        style={[styles.navItem, styleObj, focusView]}
        dt_pgid={dtConst.dt_pgid}
        dt_pg_params={FeedsUtils.getDtParams({
          rnVersion: getDeviceVisitor().getJsVersion(),
        })}
      >
        {this.state.useImage ? this.renderImageTab(itemWidth) : this.renderTextTab(itemWidth)}
        {this.renderRedDot(extStyles)}
      </View>
    );
  }

  public renderImageTab = (itemWidth: number): JSX.Element | null => {
    const extStyles = DEFAULT_FEEDS_STYLE.data.extTabList;
    const { tabInfo, curTabId } = this.props;

    if (!tabInfo?.tabIcons?.length) return null;

    const url = this.getTabImageUrl(tabInfo.tabIcons, tabInfo.tabId === curTabId);

    // 没有 url 则直接回退到文字
    if (!url) return this.renderTextTab(itemWidth);

    // android做异形非选中态marginTop微调
    if (getDeviceVisitor().isAdr()) {
      (extStyles.specialImage as any).marginTop = 1;
    }

    const imgStyle = tabInfo.tabId === curTabId ? extStyles.specialFocusImage : extStyles.specialImage;

    return (
      <Image
        source={{ uri: url }}
        style={[imgStyle, { width: itemWidth }]}
        noPicMode={{ enable: false }}
        nightMode={{ enable: false }}
        onError={this.onError}
        accessibilityLabel={tabInfo.tabTitle || '标签'}
        accessible
        reportData={{ sourceFrom: 'tabImage' }}
      />
    );
  };

  public renderTextTab = (itemWidth: number): JSX.Element => {
    const styles = DEFAULT_FEEDS_STYLE.data.tabList;
    const { tabInfo, curTabId } = this.props;

    const focusText = tabInfo.tabId === curTabId ? styles.focusText : null;

    return (
      <Text
        style={[
          styles.itemText,
          focusText,
          {
            width: itemWidth,
          },
          getTabpageBackgroundManager().getTabTextColors(),
        ]}
        numberOfLines={1}
      >
        {tabInfo.tabTitle || tabInfo.specialTitle }
      </Text>
    );
  };

  public renderRedDot = (extStyles: Record<string, any>): JSX.Element | null => {
    const { redDotNum } = this.props;

    if (!redDotNum) return null;

    let redDotWidth = extStyles.redDotWidth1;
    let rightPos = extStyles.redRightPos;
    if (redDotNum >= 10) {
      redDotWidth = extStyles.redDotWidth2;
      rightPos -= 1;
    }

    return (
      <View
        style={[
          extStyles.redDot,
          {
            right: rightPos,
            width: redDotWidth,
          },
        ]}
      >
        {redDotNum > 99 ? (
          <Image
            style={extStyles.redEllipsis}
            source={{ uri: FeedsIcon.ellipsis }}
            noPicMode={{ enable: false }}
          />
        ) : (
          <Text
            style={[
              extStyles.redDotNum,
              {
                marginTop: Platform.OS === 'android' ? -1 : 0,
              },
            ]}
            numberOfLines={1}
          >
            {redDotNum}
          </Text>
        )}
      </View>
    );
  };

  public getTabImageUrl = (tabIcons: TabIconItem[], selected: boolean): string => {
    const icon = tabIcons.find((icon) => {
      const scene = icon.scene as unknown as SkinModelType;
      return scene === FeedsTheme.skinMode;
    });

    if (!icon) return '';

    return selected ? icon?.selectedIconUrl : icon?.normalIconUrl;
  };

  private onPrefetchSuccess = (tabId: TabId): void => {
    this.prefetchSuccess = true;

    this.setState({
      useImage: true,
    });

    if (!iconPrefetchedTabs.includes(tabId)) {
      iconPrefetchedTabs.push(tabId);
    }
  };

  private prefetchImages = async (images: string[]): Promise<boolean[]> => {
    try {
      const result = await Promise.all(images.map(imageUrl => Image.getSize(imageUrl)));
      return result.map(size => !!size?.width);
    } catch (err) {
      addKeylink(`${this.props.tabInfo.tabId}-tabicon-prefetcherror`, KeylinkCmd.PR_ERROR_SUM);
      logError(JSON.stringify({ err, images }), `${TAG}.prefetchImages`);
      return [false];
    }
  };

  private onError = (err: any): void => {
    this.setState({
      useImage: false,
    });

    addKeylink(`${this.props.tabInfo.tabId}-tabicon-loaderror`, KeylinkCmd.PR_ERROR_SUM);
    logError(err, `${TAG}.onError`);
  };
}
