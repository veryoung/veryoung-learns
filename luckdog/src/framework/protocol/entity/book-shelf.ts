import { BaseBook } from './books';
import { Tag } from './tag';

export class BookInShelf extends BaseBook {
  /** 左上角标 */
  public leftTag: Tag = new Tag();
  /** 右上角标 */
  public rightTag: Tag = new Tag();
  /** 阅读进度 */
  public readProgress = '';
  /** 是否真实书籍 */
  public bookType = true;
}

export class CollectionInShelf  {
  /** 图片地址(最多四张) */
  public picUrls?: string[] = [];
  /** 跳转地址 */
  public jumpUrl?: string;
  /** 藏书总数 */
  public total = 0;
}
