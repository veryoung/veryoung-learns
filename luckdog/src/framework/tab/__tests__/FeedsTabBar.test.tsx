import { TabIconType } from '@/entity';
import { DynamicTabType } from '@/luckbox';
import React from 'react';
import renderer from 'react-test-renderer';
import FeedsTabBar from '../FeedsTabBar';

const tabInfo = tabIcons => ({
  tabId: 190,
  tabTitle: '故事馆',
  tabType: DynamicTabType.webview,
  webViewUrl: 'https://ugssr-server.html5.qq.com/novel-video-tab?resourcefrom=webview&can_media_play=1&canhorizontalscroll=1&ch=&tabfrom=bottom&novelfeedsversion=8345&strageid=969631_2367983_2253370_2253367_2178766_2075047_1980613_1979053_1958206_1957471_1810049_1715607_1584413_2337220',
  hippyViewName: 'storyView',
  feedsViewId: '-1',
  tabIcons: tabIcons,
  specialTitle: '故事馆',
  extinfo: {},
  status: 1,
  priority: 3,
  experiment: {
    experKey: 'showUgVideoTab',
    experVal: [1],
  },
  tabBackground: {},
  limitVersion: {},
  tabRule: {
    jumpRule: {
      startTime: '',
      endTime: '',
      validDays: '1',
      clearDays: '',
      clearTime: '',
    },
  },
});

const tabIcons = {
  textTab: [],
  iconTab: [
    {
      scene: 0,
      normalIconUrl: 'https://qbnovel.qq.com/static/ug_video_tap',
      selectedIconUrl: 'https://qbnovel.qq.com/static/ug_video_tab_on',
      iconType: TabIconType.PIC_TWO,
    }, {
      scene: 1,
      normalIconUrl: 'https://qbnovel.qq.com/static/ug_video_tapD',
      selectedIconUrl: 'https://qbnovel.qq.com/static/ug_video_tab_onD',
      iconType: TabIconType.PIC_TWO,
    },
  ],
  // 图片请求失败
  iconFetchErrorTab: [
    {
      scene: 0,
      normalIconUrl: 'https://qbnovel.qq.com/an_error_url',
      selectedIconUrl: 'https://qbnovel.qq.com/static/an_error_url',
      iconType: TabIconType.PIC_TWO,
    }, {
      scene: 1,
      normalIconUrl: 'https://qbnovel.qq.com/static/an_error_url',
      selectedIconUrl: 'https://qbnovel.qq.com/static/an_error_url',
      iconType: TabIconType.PIC_TWO,
    },
  ],
};

describe('<FeedsTabBar/>', () => {
  describe('render testing', () => {
    it('render text tab', () => {
      const props = {
        curTabId: 181,
        left: true,
        redDotNum: 0,
        right: false,
        tabInfo: tabInfo(tabIcons.textTab),
        tabsCount: 6,
      };
      const tree = renderer.create(<FeedsTabBar {...props} />);
      expect(tree).toMatchSnapshot();
    });

    it('render icon tab', () => {
      const props = {
        curTabId: 181,
        left: true,
        redDotNum: 0,
        right: false,
        tabInfo: tabInfo(tabIcons.iconTab),
        tabsCount: 6,
      };
      const tree = renderer.create(<FeedsTabBar {...props} />);
      expect(tree).toMatchSnapshot();
    });

    it('render icon tab when icon fetch error', () => {
      const props = {
        curTabId: 181,
        left: true,
        redDotNum: 0,
        right: false,
        tabInfo: tabInfo(tabIcons.iconFetchErrorTab),
        tabsCount: 6,
      };
      const tree = renderer.create(<FeedsTabBar {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
