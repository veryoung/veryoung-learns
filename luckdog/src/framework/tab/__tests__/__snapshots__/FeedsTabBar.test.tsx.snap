// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedsTabBar/> render testing render icon tab 1`] = `
<div
  dt_pg_params="rnVersion=8352&tab_id=112"
  dt_pgid="14"
  nativeName="View"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColors": Array [
          "transparent",
        ],
        "justifyContent": "center",
        "overflow": "visible",
      },
      Object {
        "paddingLeft": 12,
        "paddingRight": 7.5,
      },
      null,
    ]
  }
>
  <img
    accessibilityLabel="故事馆"
    accessible={true}
    alt=""
    nativeName="Image"
    nightMode={
      Object {
        "enable": false,
      }
    }
    noPicMode={
      Object {
        "enable": false,
      }
    }
    onError={[Function]}
    reportData={
      Object {
        "sourceFrom": "tabImage",
      }
    }
    src="https://qbnovel.qq.com/static/ug_video_tap"
    style={
      Object {
        "0": Object {
          "backgroundColors": Array [
            "transparent",
            "transparent",
            "transparent",
            "transparent",
          ],
          "height": 40,
          "marginTop": 1,
          "resizeMode": "contain",
        },
        "1": Object {
          "width": 53,
        },
      }
    }
  />
</div>
`;

exports[`<FeedsTabBar/> render testing render icon tab when icon fetch error 1`] = `
<div
  dt_pg_params="rnVersion=8352&tab_id=112"
  dt_pgid="14"
  nativeName="View"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColors": Array [
          "transparent",
        ],
        "justifyContent": "center",
        "overflow": "visible",
      },
      Object {
        "paddingLeft": 12,
        "paddingRight": 7.5,
      },
      null,
    ]
  }
>
  <img
    accessibilityLabel="故事馆"
    accessible={true}
    alt=""
    nativeName="Image"
    nightMode={
      Object {
        "enable": false,
      }
    }
    noPicMode={
      Object {
        "enable": false,
      }
    }
    onError={[Function]}
    reportData={
      Object {
        "sourceFrom": "tabImage",
      }
    }
    src="https://qbnovel.qq.com/an_error_url"
    style={
      Object {
        "0": Object {
          "backgroundColors": Array [
            "transparent",
            "transparent",
            "transparent",
            "transparent",
          ],
          "height": 40,
          "marginTop": 1,
          "resizeMode": "contain",
        },
        "1": Object {
          "width": 53,
        },
      }
    }
  />
</div>
`;

exports[`<FeedsTabBar/> render testing render text tab 1`] = `
<div
  dt_pg_params="rnVersion=8352&tab_id=112"
  dt_pgid="14"
  nativeName="View"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColors": Array [
          "transparent",
        ],
        "justifyContent": "center",
        "overflow": "visible",
      },
      Object {
        "paddingLeft": 12,
        "paddingRight": 7.5,
      },
      null,
    ]
  }
>
  <p
    nativeName="Text"
    numberOfLines={1}
    style={
      Array [
        Object {
          "colors": Array [
            "#242424",
            "#686d74",
            "#242424",
            "white",
          ],
          "fontSize": 16,
          "fontWeight": "normal",
          "opacity": 0.8,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        null,
        Object {
          "width": 49,
        },
        Object {
          "colors": Array [
            "#242424",
            "#686d74",
            "#242424",
            "white",
          ],
        },
      ]
    }
    text="故事馆"
  >
    故事馆
  </p>
</div>
`;
